(RK-307) Ignore branches for deployment with specified prefixes

andersonmills opened the pull request:
This is a partial implementation of feature request #574

This PR implements ignoring branches based on a prefix list provided in the r10k.yaml configuration file for a given source. There is a new setting option for a source called ignore_branch_prefixes:. If specified, then only branches that do not start with a string in the list will be deployed into environments. If ignore_branch_prefixes: is not specified, then all branches will be deployed. This makes it a backwards compatible change.

Please see the updated documentation for configuration in this PR for examples.

Comments and reviews appreciated.

The original work for this came from @nmaludy and is adapted from #757

nmaludy committed:
Initial implementation of branch filtering

(cherry picked from commit 7b7f0efb916ad78add918d14f88a9fe389b10054)

nmaludy committed:
Initial implementation of branch filtering

(cherry picked from commit 7b7f0efb916ad78add918d14f88a9fe389b10054)

nmaludy committed:
Fixed regex documentation

(cherry picked from commit e4df9fd3dc4db5957bae231707f39715b8008d24)

nmaludy committed:
Fixed regex documentation

(cherry picked from commit e4df9fd3dc4db5957bae231707f39715b8008d24)

andersonmills committed:
(RK-305) Add --no-force to deploy action.

This commit adds a --no-force flag to the deploy action. The
default of the r10k deploy action is to deploy all environments
regardless of local modifications in the modules. The
--no-force flag, the user can choose to have the deploy skip
environments where local modifications have occurred to the modules.

andersonmills committed:
(RK-306) Remove dependency on semantic_puppet.

This commit uses a helper method in puppet_forge for version validation.
This change removes the direct dependency of r10k on semantic_puppet.

nmaludy committed:
Changes to branch filtering based on review comments

(cherry picked from commit e634e6c39c49619046ca83b29ad165ea0781b424)

nmaludy committed:
Changes to branch filtering based on review comments

(cherry picked from commit e634e6c39c49619046ca83b29ad165ea0781b424)

haus commented on doc/dynamic-environments/configuration.mkd:
```diff
@@ -356,6 +356,29 @@ sources:
+sources:
+  mysource:
+    basedir: '/etc/puppet/environments'
+    branch_filter: '^(prod)|(qa)$'
```
You shouldn't need 2 groups here. `'^(prod|qa)$'` should also work?

nmaludy replied:
Yes, but it also doesn't matter. We're not actually using the group information so having 1 or 2 groups makes no difference.

haus commented on doc/dynamic-environments/configuration.mkd:
```diff
@@ -454,3 +477,46 @@ This will create the following directory structure:
+  app1_data:
+    remote: 'git://git-server.site/my-org/app1-hieradata'
+    basedir: '/etc/puppet/hieradata'
+    branch_filter: '^(prod-)\w+$'
```
You shouldn't need a grouping here at all?

nmaludy replied:
No, it doesn't matter in terms of functionality, but does help to logically separate out the prefix from the "extra stuff at the end".

haus commented on doc/dynamic-environments/configuration.mkd:
```diff
@@ -454,3 +477,46 @@ This will create the following directory structure:
+  app1_data:
+    remote: 'git://git-server.site/my-org/app1-hieradata'
+    basedir: '/etc/puppet/hieradata'
+    branch_filter: '^(master)|(production)$'
```
This can be `'^(master|production)$'`.

nmaludy replied:
Yes, but it also doesn't matter. We're not actually using the group information so having 1 or 2 groups makes no difference.

haus commented on doc/dynamic-environments/configuration.mkd:
```diff
@@ -454,3 +477,46 @@ This will create the following directory structure:
+
+#### Filtering branches - Exclude branches with a prefix
+
+If you would like to filter out all branches starting with and including "dev-".
```
Excluding branches this way feels a bit awkward. I wonder if having two filters, one for include and one for exclude would be better? That way you could write positive regexes in both cases, which is usually easier. I think in that case you would add `'^dev-.*'` to an exclude regex and any matches would be excluded. Not sure if that is better or not.

nmaludy replied:
What happens if they specify both filters, then which filter applies?

I would be more inclined to implement an additional option branch_filter_type that can either be include or exclude. Then the regex will be interpreted as either inclusive or exclusive.

andersonmills replied:
I'm a big fan of the simplicity of having one filter and using the regex negation. I expect that users getting to this level of usage will be able to follow the examples or be able to look it up quickly.

haus replied:
I feel like adding a setting that is a regex is probably a sign that we're taking the wrong approach to solving the problem. Do users actually need full regex matching, or would a simple allow/reject on a prefix or lists of allowed/denied environments be adequate?
Regexes tend to be simple in implementation, but not in actual usage, and debugging them can be challenging.

andersonmills replied:
I can see users wanting to both be able to use a list of denied/allowed environments, and a list of denied/allowed prefixes. As I start thinking about how to document how this works for users, it gets ...uh... complicated. Do you have an configuration spec in mind?

haus replied:
I would probably start with something like `ignore_branches` or `ignore_branch_prefixes` as a yaml array. I suspect that having a way to reject environments with a given prefix would be sufficient for most users.

haus replied:
`ignore_branch_prefixes` could also be a list of environments to ignore.

andersonmills replied:
I like it.

haus commented on lib/r10k/source/git.rb:
```diff
@@ -113,7 +119,17 @@ def desired_contents
+    branches = @cache.branches
+    if !@branch_filter.nil? && !@branch_filter.empty?
+      branches = branches.select do |branch|
+        result = branch[/#{@branch_filter}/]
```
This regex match syntax is fairly weird for ruby. I would expect something more like:

```
filter = Regexp.new(@branch_filter)
...
result = filter.match(branch)
...
```

andersonmills replied:
Making the change.

haus commented on lib/r10k/source/git.rb:
```diff
@@ -113,7 +119,17 @@ def desired_contents
     opts = {:prefix => @prefix, :invalid => @invalid_branches, :source => @name}
-    @cache.branches.map do |branch|
+    branches = @cache.branches
+    if !@branch_filter.nil? && !@branch_filter.empty?
```
The double negation here is funky. I would recommend a partially positive test instead such as...

`if @branch_filter && !@branch_filter.empty?`
or
`if @branch_filter && @branch_filter.size > 0`

andersonmills replied:
Opting for the former suggestion.

haus commented on lib/r10k/source/git.rb:
```diff
@@ -113,7 +119,17 @@ def desired_contents
-    @cache.branches.map do |branch|
+    branches = @cache.branches
+    if !@branch_filter.nil? && !@branch_filter.empty?
+      branches = branches.select do |branch|
```
The select call could be lifted into its own method and then unit tested on its own. `filter_branches` or something of the sort.

andersonmills replied:
Workin' on that.

andersonmills replied:
Well, I lifted it out, but it's a private method, as is the branch_names method. Should we be unit testing the private methods? I can, but I thought that was considered bad testing methodology.

andersonmills committed:
(RK-307) Update the :branch_filter: implementation.

This commit addresses PR comments around the implementation of
the :branch_filter: setting. It changes to easier to read and more
ruby-idiomatic code.

andersonmills committed:
(RK-307) Update the :branch_filter: implementation.

This commit addresses PR comments around the implementation of
the :branch_filter: setting. It changes to easier to read and more
ruby-idiomatic code.

andersonmills committed:
Merge pull request #775 from andersonmills/RK-306-version_validation_through_forge-ruby

(RK-306) Remove dependency on semantic_puppet.

andersonmills committed:
(RK-307) Ignore branches for deployment.

This commit changes from using a regex for branch matching to using a
list of strings as branch prefixes to ignore. Using a regex string as a
setting was considered too difficult to debug in practice, and this was
a better alternative.

andersonmills committed:
(RK-307) Ignore branches for deployment.

This commit changes from using a regex for branch matching to using a
list of strings as branch prefixes to ignore. Using a regex string as a
setting was considered too difficult to debug in practice, and this was
a better alternative.

haus commented on lib/r10k/source/git.rb:
```diff
@@ -111,9 +117,25 @@ def desired_contents
 
+  def filter_branches(branches, ignore_prefixes)
+    filter = Regexp.new("^(#{ignore_prefixes.join('|')}).?")
+    branches = branches.select do |branch|
```
i think if you use `reject` (https://ruby-doc.org/core-2.4.2/Enumerable.html#method-i-reject) instead of `select` here you can avoid the negation of `result` below.

andersonmills replied:
Changing.

haus commented on lib/r10k/source/git.rb:
```diff
@@ -111,9 +117,25 @@ def desired_contents
   private
 
+  def filter_branches(branches, ignore_prefixes)
+    filter = Regexp.new("^(#{ignore_prefixes.join('|')}).?")
```
What is the `.?` doing in the regex? I assume it is an optional `any character`, but I'm not sure why it is needed.

andersonmills replied:
It is not. Removing.

haus commented on lib/r10k/source/svn.rb:
```diff
@@ -106,6 +112,13 @@ def names_and_paths
 
     branches << [R10K::Environment::Name.new('production', opts), "#{@remote}/trunk"]
     @svn_remote.branches.each do |branch|
+      if !@branch_filter.nil? && !@branch_filter.empty?
```
I think the same change as was made to git could be applied here. (Both the conditional change and the ignore_branch_prefix change).

andersonmills replied:
D'oh! It's been so long since I paid attention to svn in r10k. Adding.

andersonmills committed:
(RK-307) Add svn branch ignoring and extra unit tests.

This commit adds ignore_branch_prefixes setings to svn sources and adds
unit tests for both that and branch filtering on both git and svn
sources. Previous commits in the RK-307 changes had left these out.

andersonmills committed:
(RK-307) Add svn branch ignoring and extra unit tests.

This commit adds ignore_branch_prefixes setings to svn sources and adds
unit tests for both that and branch filtering on both git and svn
sources. Previous commits in the RK-307 changes had left these out.

andersonmills committed:
(RK-264) Change "update_force" to "force"

This commit changes the --update_force flag to just a --force flag.
This improves the UX and puts it in line with the final implementation
of RK-305.

haus commented on lib/r10k/source/git.rb:
```diff
@@ -109,11 +115,27 @@ def desired_contents
   end
 
+  def filter_branches(branches, ignore_prefixes)
+    filter = Regexp.new("^(#{ignore_prefixes.join('|')})")
```
This looks pretty good. I think `Regexp.new("^#{Regexp.union(ignore_prefixes)}")` is equivalent, but I'm not sure it's any cleaner.

andersonmills replied:
It's maybe a little clearer (maybe).

haus commented on spec/unit/source/git_spec.rb:
```diff
@@ -62,6 +62,38 @@
     end
   end
+
+  describe "generating environments without the ignore_branch_prefixes" do
```
I think this test description is a little funky. When i first read it I presumes it would be a test without using the ignore_branch_prefixes setting. Maybe `describe "generate_environments respects ignore_branch_prefixes setting"`?

andersonmills replied:
Yeah, this has been through a few iterations to get here.

andersonmills replied:
Yeah this went through too many editing iterations.

haus commented on spec/unit/source/git_spec.rb:
```diff
@@ -62,6 +62,38 @@
+  end
+
+  describe "filtering branches with ignore prefixes" do
+    let(:branches) { ['master', 'development', 'production', 'dev-test', 'dev', 'test_2'] }
```
I think a test branch with dev or test at the end or the middle of a valid branch name would be good.

andersonmills replied:
Good idea.

andersonmills replied:
Adding this caused me to notice that "dev-test" is not a valid environment name, anyway.

haus commented on spec/unit/source/svn_spec.rb:
```diff
@@ -61,6 +61,33 @@
     end
   end
+
+  describe "generating environments with branch filter exclude prefix" do
```
This describe should probably be the same as whatever you end up with for git.

haus commented on spec/unit/source/svn_spec.rb:
```diff
@@ -61,6 +61,33 @@
+  end
+
+  describe "filtering branches with ignore prefixes" do
+    let(:branches) { ['master', 'development', 'production', 'dev-test', 'dev', 'test_2'] }
```
Same comment as in the git about having a test case with dev/test in the middle or end of a valid branch name.

andersonmills committed:
(RK-307) Improve exclusion tests and regex.

This commit improves the tests for ignore_branch_prefixes and the
readability of the regex created for filtering.

andersonmills committed:
(RK-307) Improve exclusion tests and regex.

This commit improves the tests for ignore_branch_prefixes and the
readability of the regex created for filtering.

haus reviewed (approved):
Sweet. Looks great.

andersonmills committed:
(RK-305) Use @force instead of @no_force for logic ease.

This commit changes to using a @force variable to represent the
opposite of the --no-force flag in order to make logic easier to read in
deploy/environment and deploy/module.

andersonmills committed:
Merge pull request #780 from andersonmills/RK-264-change_update_force_to_force

(RK-264) Change "update_force" to "force"

andersonmills committed:
(RK-305) Update allowed options for dash in no-force.

This commit updates the allowed options for deploy/environment and
deploy/module for the dash in the no-force flag.

andersonmills committed:
(RK-305) Fix string-symbol hash notation for ruby < 2.2.0.

This commit fixes the hash notation for string symbols to be compatible
with rubies earlier than 2.2.0. The short hand of { 'key': value } does
not work. { :'key' => value } must be used.

andersonmills committed:
Merge pull request #774 from andersonmills/RK-305-preserve_force_flag

(RK-305) Add --no-force to deploy action.

andersonmills committed:
(maint) Merge branch '2.5.x' into 2.6.x

Commit:
Merge pull request #781 from andersonmills/merge_up_2.5.x_into_2.6.x

(maint) Merge up 2.5.x into 2.6.x

andersonmills committed:
Merge branch 'myPR757_for_2.6.x' of github.com:andersonmills/r10k into myPR757_for_2.6.x
